### PR TITLE
Add native geolocation as main geoloc strategy.

### PR DESCRIPTION
**Pull request by lemavri**
This is a rebased version of the location branch.

TODO

- [x] Fallback to alternative geoloc strategy (not GeoIP, which sucks)
- [x] Big G geoloc specs
- [x] Native Geoloc specs
- [x] GeoIP specs
- [x] Preliminary build + QA

---

devlog: 

whereami is not prompting user to give permissions to access the location when ran from Prey. Perhaps we need to add it manually by running `defaults write /var/db/locationd/clients.plist` in the post_install script.

### CODE REVIEW

File: lib/agent/helpers.js [hunk 1]
```diff
@@ -21,24 +21,41 @@ helpers.run_via_service = function(){
...
+    return false;
   }
+  return true;
+}
 
-  return semver.gt(first, second);
-};
+function handle_version_error(method_name, versions) {
+  var err_msg = "Cannot run" + method_name + ". Invalid versions: ";
+
```
Comment (raliste) [inline lib/agent/helpers.js]: Unnecessary lines 55 and 58

File: lib/agent/providers/geo/darwin/index.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    child_process = require('child_process'),
+    common = require('../../../common'),
+    system = common.system,
+    gte = common.helpers.is_greater_or_equal;
+
+// whereami bin from https://github.com/robmathers/WhereAmI
+var bin = join(__dirname, 'bin', 'whereami');
+
+var get_using_corelocation = function(cb) {
+
```
Comment (raliste) [inline lib/agent/providers/geo/darwin/index.js]: Idem

File: lib/agent/providers/geo/darwin/index.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      res.accuracy = match[1];
+    } else if (match = str.match(/Horizontal Accuracy.*: (.+)/)) {
+      res.accuracy = match[1];
+      res.vertical_accuracy = str.match(/Vertical Accuracy.*: (.+)/)[1];
+      res.altitude = str.match(/Altitude.*: (.+)/)[1];
+    }
+
+    cb(null, res);
+  })
+
```
Comment (raliste) [inline lib/agent/providers/geo/darwin/index.js]: Idem

File: lib/agent/providers/geo/darwin/index.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+exports.get_location = function(cb) {
+
```
Comment (raliste) [inline lib/agent/providers/geo/darwin/index.js]: Idem

File: lib/agent/providers/geo/darwin/index.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  system.get_os_version(function(err, version) {
+    if (version && gte(version, "10.6.0"))
+      return get_using_corelocation(cb);
+
+    cb(new Error('CoreLocation not suppored in OSX ' + version))
+  });
+
```
Comment (raliste) [inline lib/agent/providers/geo/darwin/index.js]: Idem

File: lib/agent/providers/geo/linux/index.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+function get_command_one(provider) {
+  var bin     = 'dbus-send',
+      service = 'org.freedesktop.Geoclue.Providers.' + provider,
+      path    = '/org/freedesktop/Geoclue/Providers/' + provider,
+      command = 'org.freedesktop.Geoclue.Position.GetPosition';
+
+  return [bin, '--print-reply', '--dest=' + service, path, command].join(' ');
+}
+
```
Comment (raliste) [inline lib/agent/providers/geo/linux/index.js]: Extra line

File: lib/agent/providers/geo/strategies.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
+"use strict"
+
+var join = require('path').join,
+  needle = require('needle'),
```
Comment (raliste) [inline lib/agent/providers/geo/strategies.js]: Wrong alignment

File: lib/agent/providers/geo/strategies.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+  get_agent = require('random-ua').generate,
+  platform = require(join(__dirname, process.platform)),
+  providers = require('./../../providers'),
+  logger = require('../../common').logger.prefix('geo');
+
+function geoip(cb) {
+
```
Comment (raliste) [inline lib/agent/providers/geo/strategies.js]: Extra line

File: lib/agent/providers/geo/strategies.js [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+      lng: parseFloat(body.loc.split(',')[1]),
+      method: 'geoip'
+    }
+
+    cb(null, res);
+  })
+}
+
+function google(cb) {
+
```
Comment (raliste) [inline lib/agent/providers/geo/strategies.js]: Extra lines, not consistent with previous style. Review extra lines moving down as well.

File: test/lib/agent/providers/geo/geoip_strategy.js [hunk 10]
```diff
@@ -0,0 +1,122 @@
+var helpers = require('./../../../../helpers'),
+  should = require('should'),
```
Comment (raliste) [inline test/lib/agent/providers/geo/geoip_strategy.js]: Wrong alignment

File: test/lib/agent/providers/geo/google_strategy.js [hunk 11]
```diff
@@ -0,0 +1,108 @@
+var helpers = require('./../../../../helpers'),
+  should = require('should'),
+  google_strat = helpers.load('providers/geo/strategies').google,
+  location_response = require('../fixtures/google_location_response');
+
+describe('location', function() {
+
+  describe('when access points return error', function() {
+
```
Comment (raliste) [inline test/lib/agent/providers/geo/google_strategy.js]: Missing specs

**Comment by lemavri** [comment]
@raliste done!

**Comment by raliste** [comment]
Nice :+1:

**Pull request merged**